src/role_scout.js

```diff
@@ -13,153 +13,226 @@ roles.scout.settings = {
-      //       let roomNext = rooms[direction];
-      //       if (haveNotSeen(creep, roomNext)) {
-      //         creep.memory.search.levels[creep.memory.search.level + 1].push(roomNext);
-      //         creep.memory.search.target = roomNext;
-      //       }
-      //     }
-      //   }
-      //   creep.memory.search.level++;
-      // }
-      return false;
+roles.scout.goFurther = function(creep) {
+  creep.memory.search.levels[creep.memory.search.level + 1] = [];
+  let target = roles.scout.lookAround(creep, creep.memory.search.level + 1, true);
+  if (target) {
+    if (!Memory.rooms[target]) {
+      Memory.rooms[target] = {};
     }
+    Memory.rooms[target].scoutSeen = creep.name;
+  }
+
+  if (creep.memory.search.levels[creep.memory.search.level + 1].length === 0) {
+    return false;
+  } // debug line
+  creep.memory.search.level++;
+};
+
+roles.scout.switchTarget = function(creep) {
+  if (creep.memory.scoutSkip) {
+    creep.memory.skip.push(creep.memory.search.target);
+    delete creep.memory.scoutSkip;
+  } else if (creep.room.name === creep.memory.search.target) {
+    creep.memory.search.seen.push(creep.room.name);
+  }
+  if (!roles.scout.lookAround(creep, 0, false, true) &&
+  !roles.scout.setNewTarget(creep)) {
+    roles.scout.goFurther(creep);
+  }
+};
 
-    if (creep.memory.last && creep.memory.last.pos3 && creep.pos.roomName !== creep.memory.last.pos3.roomName) {
-      creep.moveTo(25, 25);
-      return true;
+roles.scout.move = function(creep) {
+  let costMatrix = function(roomName) {
+    if (!Memory.rooms[roomName] || (Memory.rooms[roomName].lastSeen < (Game.time - 5))) {
+      return creep.room.getCostMatrixCallback(creep.targetPosObject, true, false, true);
     }
+    return false;
+  };
 
-    if (creep.isStuck()) {
-      creep.moveRandom();
-      creep.say('ImStuck', true);
-      creep.log('Scout Stuck, Randomly Moving: ' + JSON.stringify(creep.memory.last) + ' ' + JSON.stringify(creep.isStuck()));
-      return true;
+  let search = PathFinder.search(
+    creep.pos, {
+      pos: creep.targetPosObject,
+      range: 20
+    }, {
+      roomCallback: costMatrix
     }
+  );
+
+  if (config.visualizer.enabled && config.visualizer.showPathSearches) {
+    visualizer.showSearch(search);
+  }
 
-    if (search.path.length === 0 || (creep.inBase() && creep.room.memory.misplacedSpawn)) {
-      creep.say('hello', true);
-      //       creep.log(creep.pos + ' ' + targetPosObject + ' ' + JSON.stringify(search));
-      if (creep.isStuck() && onBorder(creep)) {
-        creep.say('imstuck at the border', true);
-        if (config.room.scoutSkipWhenStuck) {
-          creep.say('skipping', true);
-          creep.memory.scoutSkip = true;
-          delete creep.memory.last; // Delete to reset stuckness.
-        }
+  if (search.path.length === 0 || (creep.inBase() && creep.room.memory.misplacedSpawn)) {
+    if (creep.isStuck() && roles.scout.onBorder(creep)) {
```

Comment:
`creep.pos.isBorder()`?

Reply:
yup should be better ^^ onBorder was already here as i didn't look if there was a creep proto.